sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -279,6 +296,49 @@ class ArtifactManager(session: SparkSession) extends Logging {
     loader
   }
 
+  private[sql] def clone(newSession: SparkSession): ArtifactManager = {
+    val sparkContext = session.sparkContext
+    sparkContext.synchronized {
```

Comment:
What are you trying to achieve with this lock? It seems a bit coarse :)...

Reply:
This lock is to avoid concurrently calling `sparkContext.addJar` because it's not thread-safe.
I think it's possible to reduce the lock scope to the "re-register" part.

Reply:
Oops no, they are thread-safe. I'll remove this lock.

Reply:
You put it back?

Reply:
How about locking on a new object in the ArtifactManager companion object?

Reply:
Based on an offline discussion - the lock here is not needed due to

1. SparkContext's JAR maps are thread-safe.
2. Every `ArtifactManager.clone` call will have a fresh session UUID, so collisions are very unlikely.